<a href="https://colab.research.google.com/github/KirillSmirnov2/InnopolisStudentPublicRepository/blob/main/HW3(dataset_from_HW2)_SmirnovKK.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [105]:
import pandas as pd
import time
df1=pd.read_csv('anime.csv')

In [106]:
# @title Используем команды '.info()'  и  '.describe()' для получение информации о датасете.
df1.info()
df1.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


anime_id        rating       members
count  12294.000000  12064.000000  1.229400e+04
mean   14058.221653      6.473902  1.807134e+04
std    11455.294701      1.026746  5.482068e+04
min        1.000000      1.670000  5.000000e+00
25%     3484.250000      5.880000  2.250000e+02
50%    10260.500000      6.570000  1.550000e+03
75%    24794.500000      7.180000  9.437000e+03
max    34527.000000     10.000000  1.013917e+06

In [107]:
# @title Определим общий список жанров в датасете
n_rows=len(df1['genre'])
print('число строк:  ',n_rows)
genre_list=[]
count_genres=0
for i in range(n_rows): #n_rows#
  S1=[]
  if type(df1['genre'].loc[i]) == str:
    S1=df1['genre'].loc[i].split(',')
    P=[]
    for j in S1:
      P=j.replace(' ','')
      if genre_list.count(P)==0:
        genre_list.append(P)
  else:
    pass
print('список жанров:  ',genre_list)
print('число жанров:  ',len(genre_list))







число строк:   12294
список жанров:   ['Drama', 'Romance', 'School', 'Supernatural', 'Action', 'Adventure', 'Fantasy', 'Magic', 'Military', 'Shounen', 'Comedy', 'Historical', 'Parody', 'Samurai', 'Sci-Fi', 'Thriller', 'Sports', 'SuperPower', 'Space', 'SliceofLife', 'Mecha', 'Music', 'Mystery', 'Seinen', 'MartialArts', 'Vampire', 'Shoujo', 'Horror', 'Police', 'Psychological', 'Demons', 'Ecchi', 'Josei', 'ShounenAi', 'Game', 'Dementia', 'Harem', 'Cars', 'Kids', 'ShoujoAi', 'Hentai', 'Yaoi', 'Yuri']
число жанров:   43


**Задание 3 на основе данных из Задания 2**

**Продолжить исследование датасета, который был использован в прошлом домашнем задании. Выполнить следующие действия: **

- проверить датасет на наличие пропусков, если они есть, вывести строки с пропусками на экран;
- провести оценку корреляции между столбцами датафрейма и сделать выводы;
- произвольно применить функции сложения, нахождения среднего значения, медианы, минимума и максимума, стандартного отклонения для случайной выборки из датасета;
- применить операции сравнения.


In [108]:
# @title Проверка dataset на наличие пропусков заполнения столбцов "жанра" и "типа"
df2=df1.isnull()
df_NaN_Type=df1.loc[df2['type'] == True]
df_NaN_Genre=df1.loc[df2['genre'] == True]
print('число строк без данных по типу :  ', len(df_NaN_Type) )
print('число строк без данных по жанру :  ', len(df_NaN_Genre))
print('Общее число строк  данных :  ', len(df1))
print('==============================================')
print('строки с пропусками информации по жанру :  ')
print(df_NaN_Genre[['name' ,'genre','type']])
print('==============================================')
print('строки с пропусками информации по типу :  ')
print(df_NaN_Type[['name' ,'genre','type']])
print('==============================================')

число строк без данных по типу :   25
число строк без данных по жанру :   62
Общее число строк  данных :   12294
строки с пропусками информации по жанру :  
                                                    name genre     type
2844           IS: Infinite Stratos 2 - Infinite Wedding   NaN  Special
3541                                       ViVid Strike!   NaN       TV
6040                                   Metropolis (2009)   NaN    Movie
6646                                        Match Shoujo   NaN      ONA
7018                                    Katsudou Shashin   NaN    Movie
...                                                  ...   ...      ...
11070                                    Seikaisuru Kado   NaN      NaN
11086  Tamagotchi Movie: Tanpen Himitsu no Otodoke Da...   NaN    Movie
11097                             Tsukipro The Animation   NaN       TV
11112                                  Yuyushiki Special   NaN  Special
11113                                      Zunda Ho

In [109]:
# @title Провести оценку корреляции между столбцами датафрейма и сделать выводы: Проверим корреляцию между рейтингом[rating] и количеством участников сообщества[members].
df_NaN_ratings=df1.loc[df2['rating'] == True]

print(df_NaN_ratings[['anime_id' ,'rating','members']])
print('число строк без данных по рейтингу [rating] :  ', len(df_NaN_ratings) )
print('==============================================')

df_NaN_members=df1.loc[df2['members'] == True]
print(df_NaN_members[['anime_id' ,'rating','members']])
print('число строк без данных по числу участников [members] :  ', len(df_NaN_members) )
print('==============================================')
df_ratings_OK=df1.loc[df2['rating'] == False]
df_corr=df_ratings_OK.corr()
print(len(df_ratings_OK))
print('проверка строк:  ',n_rows,' - ',len(df_ratings_OK),'=',(n_rows - len(df_ratings_OK)), '|  ' , len(df_NaN_ratings))
print(df_corr)


       anime_id  rating  members
8968      34502     NaN       32
9657      34309     NaN      129
10896     34096     NaN    13383
10897     34134     NaN    90706
10898     30484     NaN    60999
...         ...     ...      ...
12274     34492     NaN      392
12279     34491     NaN       79
12280     34312     NaN      240
12282     34388     NaN      195
12285     34399     NaN      485

[230 rows x 3 columns]
число строк без данных по рейтингу [rating] :   230
Empty DataFrame
Columns: [anime_id, rating, members]
Index: []
число строк без данных по числу участников [members] :   0
12064
проверка строк:   12294  -  12064 = 230 |   230
          anime_id    rating   members
anime_id  1.000000 -0.284625 -0.075500
rating   -0.284625  1.000000  0.387979
members  -0.075500  0.387979  1.000000


<ipython-input-109-b95431fff4bb>:13: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_corr=df_ratings_OK.corr()


Вывод: Слабая корреляция (r=0.387979) между числом участников и рейтингом. Рейтинг мало зависит от числа участкников сообщества.




**Произвольно применить функции сложения, нахождения среднего значения, медианы, минимума и максимума, стандартного отклонения для случайной выборки из датасета;**

Определеим число участников сообществ связанных с жанрами:
[Shounen]
[Hentai]
[Mecha]


__ --------------------------------------
Загрузим файл с записями оценок [rating.csv]
__ --------------------------------------

In [110]:
dfT=pd.read_csv('rating.csv')
dfT.info()
dfT.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7813737 entries, 0 to 7813736
Data columns (total 3 columns):
 #   Column    Dtype
---  ------    -----
 0   user_id   int64
 1   anime_id  int64
 2   rating    int64
dtypes: int64(3)
memory usage: 178.8 MB


user_id      anime_id        rating
count  7.813737e+06  7.813737e+06  7.813737e+06
mean   3.672796e+04  8.909072e+03  6.144030e+00
std    2.099795e+04  8.883950e+03  3.727800e+00
min    1.000000e+00  1.000000e+00 -1.000000e+00
25%    1.897400e+04  1.240000e+03  6.000000e+00
50%    3.679100e+04  6.213000e+03  7.000000e+00
75%    5.475700e+04  1.409300e+04  9.000000e+00
max    7.351600e+04  3.451900e+04  1.000000e+01

Теперь есть два файла по которым можно выполнять анализ

Определим списки [anime_id] по каждому жанру:  [Shounen]; [Hentai]; [Mecha]
Определим число участников сообществ [members] по [anime_id] каждому жанру по данным из [anime.csv]

определим число участников сообществ по оценкам [rating]  от [user_id] из [rating.csv]   по определенным [anime_id] (по данным из [anime.csv])


In [137]:
def Count_Members(fls,AL_G, STR_PRT, NM):
  T0 = time.time()
  print('==============================================')
  Kd=0
  Kd1=0
  print('[', end='')
  for j in range(len(AL_G)):
    if fls == 1:
      NM=NM+df1['members'].loc[df1['anime_id'] == AL_G[j]].values
    elif fls == 2:
      NM=NM+len(dfT.loc[dfT['anime_id'] == AL_G[j]])
    else:
      pass
    Kd1=j/len(AL_G)*10 // 1
    if Kd1>Kd:
      Kd = Kd1
      print('-', end='')
  T1 = time.time()
  print('-]')
  print(STR_PRT,' - ОК; ProcessTime: ', (T1-T0), ' sec')



In [138]:
AnimeIdList_genre_Shounen=[]
AnimeIdList_genre_Hentai=[]
AnimeIdList_genre_Mecha=[]
for i in range(n_rows):
  S1=[]
  if type(df1['genre'].loc[i]) == str:
    S1=df1['genre'].loc[i].split(',')
    P=[]
    for j in S1:
      P=j.replace(' ','')
      if P=='Shounen':
        AnimeIdList_genre_Shounen.append(df1['anime_id'].loc[i])
      else:
        if P=='Hentai':
          AnimeIdList_genre_Hentai.append(df1['anime_id'].loc[i])
        else:
          if P=='Mecha':
            AnimeIdList_genre_Mecha.append(df1['anime_id'].loc[i])
          else:
            pass
  else:
    pass

print('==============================================')
print('число аниме жанра [Shounen]:   ',len(AnimeIdList_genre_Shounen))
print('число аниме жанра [Hentai]:   ',len(AnimeIdList_genre_Hentai))
print('число аниме жанра [Mecha]:   ',len(AnimeIdList_genre_Mecha))
print('==============================================')
NM_v1_Shounen=0
NM_v2_Shounen=0

NM_v1_Hentai=0
NM_v2_Hentai=0

NM_v1_Mecha=0
NM_v2_Mecha=0

#efev=df1['members'].loc[df1['anime_id'] == 820]
#print(efev.values)


  #return


Count_Members(1,AnimeIdList_genre_Shounen,'NM_v1_Shounen',NM_v1_Shounen)
Count_Members(1,AnimeIdList_genre_Hentai,'NM_v1_Hentai',NM_v1_Hentai)
Count_Members(1,AnimeIdList_genre_Mecha,'NM_v1_Mecha',NM_v1_Mecha)
print('==============================================')
Count_Members(2,AnimeIdList_genre_Shounen,'NM_v2_Shounen',NM_v2_Shounen)
Count_Members(2,AnimeIdList_genre_Hentai,'NM_v2_Hentai',NM_v2_Hentai)
Count_Members(2,AnimeIdList_genre_Mecha,'NM_v2_Mecha',NM_v2_Mecha)

T0 = time.time()
print('==============================================')
Kd=0
Kd1=0
print('[', end='')
for j in range(len(AnimeIdList_genre_Shounen)):
  NM_v1_Shounen=NM_v1_Shounen+df1['members'].loc[df1['anime_id'] == AnimeIdList_genre_Shounen[j]].values
  Kd1=j/len(AnimeIdList_genre_Shounen)*10 // 1
  if Kd1>Kd:
    Kd = Kd1
    print('-', end='')
T1 = time.time()
print('-]')
print('NM_v1_Shounen - ОК; ProcessTime: ', (T1-T0), ' sec')
T0=T1

Kd=0
Kd1=0
print('[', end='')
for j in range(len(AnimeIdList_genre_Hentai)):
  NM_v1_Hentai=NM_v1_Hentai+df1['members'].loc[df1['anime_id'] == AnimeIdList_genre_Hentai[j]].values
  Kd1=j/len(AnimeIdList_genre_Hentai)*10 // 1
  if Kd1>Kd:
    Kd = Kd1
    print('-', end='')
T1 = time.time()
print('-]')
print('NM_v1_Hentai - ОК; ProcessTime: ', (T1-T0), ' sec')
T0=T1

Kd=0
Kd1=0
print('[', end='')
for j in range(len(AnimeIdList_genre_Mecha)):
  NM_v1_Mecha=NM_v1_Mecha+df1['members'].loc[df1['anime_id'] == AnimeIdList_genre_Mecha[j]].values
  Kd1=j/len(AnimeIdList_genre_Mecha)*10 // 1
  if Kd1>Kd:
    Kd = Kd1
    print('-', end='')
T1 = time.time()
print('-]')
print('NM_v1_Mecha - ОК; ProcessTime: ', (T1-T0), ' sec')
T0=T1
print('==============================================')
Kd=0
Kd1=0
print('[', end='')
for j in range(len(AnimeIdList_genre_Shounen)):
  NM_v2_Shounen=NM_v2_Shounen+len(dfT.loc[dfT['anime_id'] == AnimeIdList_genre_Shounen[j]])
  Kd1=j/len(AnimeIdList_genre_Shounen)*10 // 1
  if Kd1>Kd:
    Kd = Kd1
    print('-', end='')
T1 = time.time()
print('-]')
print('NM_v1_Shounen - ОК; ProcessTime: ', (T1-T0), ' sec')
T0=T1

Kd=0
Kd1=0
print('[', end='')
for j in range(len(AnimeIdList_genre_Hentai)):
  NM_v2_Hentai=NM_v2_Hentai+len(dfT.loc[dfT['anime_id'] == AnimeIdList_genre_Hentai[j]])
  Kd1=j/len(AnimeIdList_genre_Hentai)*10 // 1
  if Kd1>Kd:
    Kd = Kd1
    print('-', end='')
T1 = time.time()
print('-]')
print('NM_v2_Hentai - ОК; ProcessTime: ', (T1-T0), ' sec')
T0=T1

Kd=0
Kd1=0
print('[', end='')
for j in range(len(AnimeIdList_genre_Mecha)):
  NM_v2_Mecha=NM_v2_Mecha+len(dfT.loc[dfT['anime_id'] == AnimeIdList_genre_Mecha[j]])
  Kd1=j/len(AnimeIdList_genre_Mecha)*10 // 1
  if Kd1>Kd:
    Kd = Kd1
    print('-', end='')
T1 = time.time()
print('-]')
print('NM_v2_Mecha - ОК; ProcessTime: ', (T1-T0), ' sec')
T0=T1
print('==============================================')


print('================== по файлу anime.csv =======================')
print('число участников сообществ [members] по [anime_id] жанра [Shounen]:   ', NM_v1_Shounen)
print('число участников сообществ [members] по [anime_id] жанра [Hentai]:   ', NM_v1_Hentai)
print('число участников сообществ [members] по [anime_id] жанра [Mecha]:   ', NM_v1_Mecha)
print('==============================================')
print('================== по файлу rating.csv =======================')
print('число участников сообществ [members] по [anime_id] жанра [Shounen]:   ', NM_v2_Shounen)
print('число участников сообществ [members] по [anime_id] жанра [Hentai]:   ', NM_v2_Hentai)
print('число участников сообществ [members] по [anime_id] жанра [Mecha]:   ', NM_v2_Mecha)
print('==============================================')



число аниме жанра [Shounen]:    1712
число аниме жанра [Hentai]:    1141
число аниме жанра [Mecha]:    944
[----------]
NM_v1_Shounen  - ОК; ProcessTime:  0.38022565841674805  sec
[----------]
NM_v1_Hentai  - ОК; ProcessTime:  0.24108338356018066  sec
[----------]
NM_v1_Mecha  - ОК; ProcessTime:  0.2149524688720703  sec
[----------]
NM_v2_Shounen  - ОК; ProcessTime:  22.87066626548767  sec
[----------]
NM_v2_Hentai  - ОК; ProcessTime:  14.90179443359375  sec
[----------]
NM_v2_Mecha  - ОК; ProcessTime:  12.07577896118164  sec
[----------]
NM_v1_Shounen - ОК; ProcessTime:  0.35980725288391113  sec
[----------]
NM_v1_Hentai - ОК; ProcessTime:  0.2418370246887207  sec
[----------]
NM_v1_Mecha - ОК; ProcessTime:  0.19956183433532715  sec
[----------]
NM_v1_Shounen - ОК; ProcessTime:  22.081353902816772  sec
[----

KeyboardInterrupt: ignored

In [115]:
dfT.loc[dfT['anime_id'] == 820]



user_id  anime_id  rating
14129        183       820      10
18741        234       820      10
24459        288       820       9
29302        339       820      -1
55229        572       820      10
...          ...       ...     ...
7766220    73122       820       9
7767843    73132       820      10
7792556    73320       820      10
7804090    73408       820      -1
7810303    73485       820      10

[903 rows x 3 columns]

In [122]:
print(100//10)

10
